src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/permissionSelection.jsx

```diff
@@ -0,0 +1,181 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import {find, flatMap} from 'lodash';
+
+import {t} from 'app/locale';
+import {SENTRY_APP_PERMISSIONS} from 'app/constants';
+import AsyncView from 'app/views/asyncView';
+import ConsolidatedScopes from 'app/utils/consolidatedScopes';
+import SelectField from 'app/views/settings/components/forms/selectField';
+
+/**
+ * Custom form element that presents API scopes in a resource-centric way. Meaning
+ * a dropdown for each resource, that rolls up into a flat list of scopes.
+ *
+ *
+ * API Scope vs Permission
+ *
+ *    "API Scopes" are the string identifier that gates an endpoint. For example,
+ *    `project:read` or `org:admin`. They're made up of two parts:
+ *
+ *       <resource>:<access>
+ *
+ *    "Permissions" are a more user-friendly way of conveying this same information.
+ *    They're roughly the same with one exception:
+ *
+ *       - No Access
+ *       - Read
+ *       - Read & Write
+ *       - Admin
+ *
+ *    "Read & Write" actually maps to the `write` access level since `read` is
+ *    implied. Similarly, `admin` implies `read` and `write`.
+ *
+ *    This components displays things per Resource. Meaning the User selects
+ *    "Read", "Read & Write", or "Admin" for Project or Organization or etc.
+ *
+ *    == Scopes to Permissions
+ *
+ *    The first thing this component does on instantiation is take the list of API
+ *    Scopes passed via `props` and converts them to "Permissions.
+ *
+ *    So a list of scopes like the following:
+ *
+ *       ['project:read', 'project:write', 'org:admin']
+ *
+ *    will become an object that looks like:
+ *
+ *       {
+ *         'Project': 'write',
+ *         'Organization': 'admin',
+ *       }
+ *
+ *
+ * State
+ *
+ *    This component stores state like the example object from above. When the
+ *    User changes the Permission for a particular resource, it updates the
+ *    `state.permissions` object to reflect the change.
+ *
+ *
+ * Updating the Form Field Value
+ *
+ *    In addition to updating the state, when a value is changed this component
+ *    recalculates the full list of API Scopes that need to be passed to the API.
+ *
+ *    So if the User has changed Project to "Admin" and Organization to "Read & Write",
+ *    we end up with a `state.permissions` like:
+ *
+ *       {
+ *         'Project': 'admin',
+ *         'Organization': 'write',
+ *       }
+ *
+ *    From there, we calculate the full list of API Scopes. This list includes all
+ *    implied scopes, meaning the above state would result in:
+ *
+ *       ['project:read', 'project:write', 'project:admin', 'org:read', 'org:write']
+ *
+ */
+export default class PermissionSelection extends AsyncView {
+  static contextTypes = {
+    router: PropTypes.object.isRequired,
+    form: PropTypes.object,
+  };
+
```

Comment:
so I think you need to add 
```
getEndpoints() { 
  return [];
}
```
to avoid the deprecation console logs in the tests (@billyvg wanna fact check?)

Reply:
I'm not sure why the base class's method prints this out. The only time you'd see this is if you **aren't** implementing `getEndpoint`. ie. not using a deprecated method (which is correct).

I feel like it'd make more sense for `AsyncComponent` to look like:

```javascript
class AsyncComponent {
  getEndpoint() {
    return null;
  }

  getEndpoints() {
    let endpoint = this.getEndpoint();
    if (endpoint !== null) {
      console.log('getEndpoint is deprecated');
      return [['data', endpoint, this.getEndpointParams()]];
    }
    return [];
    }
}
```

That way you only see the deprecation warning if you're actually implementing `getEndpoint`.

Or maybe better yet, just don't implement `getEndpoint` in the baseclass, check `hasOwnProperty` and if it's found, write the deprecation warning?

It just seems like a shitty solution to add an empty, unnecessary method, to fix this, when this class isn't violating the deprecation.

Reply:
Actually, now that I look at this – it probably just shouldn't use `AsyncComponent`... because you know, it's not doing any network communication...

Reply:
even better ^

Reply:
I mean we should just actually deprecate `getEndpoint`